Bluetooth: L2CAP_BR: Enable Retransmission and Flow control

lylezhu2012 opened the pull request:
Enable the retransmission, flow control, enhance retransmission, and streaming feature for BR/EDR L2cap.

lylezhu2012 committed:
Bluetooth: L2CAP_BR: Enable Retransmission and Flow control

Enable signaling channel configuration for retransmission and Flow
control feature.

Send I-frame and S-frame. Support retransmission and Flow control for
sending.
Receive and handle I-frame and S-frame.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

jhedberg commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -234,12 +234,115 @@ struct bt_l2cap_le_chan {
+	/** Endpoint Link Mode.
+	 *  The value is defined as BT_L2CAP_BR_LINK_MODE_*
+	 */
+	uint8_t					mode;
```
Same here and anywhere else if you did the same. Use just spaces for alignment and tabs for indentation (or to put it in another way: the only place for tabs is in the very beginning of a line - as long as you have some non-tab character on a line there should not be any more tabs anywhere later in the line)

jhedberg replied:
Note: I know there are inconsistencies wrt this in the Bluetooth code base, but we should try to move toward the convention that I just described.

lylezhu2012 replied:
Yes. Updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -26,6 +27,84 @@ config BT_MAX_SCO_CONN
 	  Maximum number of simultaneous Bluetooth synchronous connections
 	  supported. The minimum (and default) number is 1.
 
+config BT_L2CAP_RET
+	bool "Bluetooth L2CAP restransmission mode [EXPERIMENTAL]"
+	select EXPERIMENTAL
+	help
+	  This option enables Bluetooth L2CAP restransmission mode
+
```
OK. How about to add test code to shell?

jhedberg replied:
> OK. How about to add test code to shell?

If there's some convenient and intuitive way to test this through the shell, then sure, why not.

lylezhu2012 replied:
OK. It is done.

lylezhu2012 committed:
Bluetooth: BR: Shell: Add L2CAP basic function commander

Add a sub-command set `l2cap` for command set `br`.

Move command `l2cap-register` to sub-command set. And rename it to
`register`.

Add command `connect`, `disconnect`, and `send` for command set
`l2cap`.

Remove original net buffer pool from `data_pool` to `data_rx_pool`.
Add a net buffer pool `data_tx_pool` for command `send`.

Do not wait anymore if no net buffer can be allocated from
`data_rx_pool`.

Dump all received data in L2CAP data received callback.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -234,12 +234,115 @@ struct bt_l2cap_le_chan {
+/** Frame Check Sequence type. No FCS. */
+#define BT_L2CAP_BR_FCS_NO			 0x00
+/** Frame Check Sequence type. 16-bit FCS. */
+#define BT_L2CAP_BR_FCS_16BIT		 0x01
```
Yes. Updated

lylezhu2012 committed:
Bluetooth: BR: Shell: L2CAP mode support

Add more parameters to command `connect` and `register`, including mode,
mode_optional, extended_control, and hold_credit.

Add command `credits` to give the rx credit.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: L2CAP_BR: Add mode optional support

Add mode optional support for BR l2cap connect initiator role.

If `chan->rx.optional` is true, set the mode to basic mode instead of
return error code `-ENOTSUP`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

jhedberg commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -26,6 +27,84 @@ config BT_MAX_SCO_CONN
 	  Maximum number of simultaneous Bluetooth synchronous connections
 	  supported. The minimum (and default) number is 1.
 
+config BT_L2CAP_RET
+	bool "Bluetooth L2CAP retransmission mode [EXPERIMENTAL]"
+	select EXPERIMENTAL
+	help
+	  This option enables Bluetooth L2CAP retransmission mode
+
+config BT_L2CAP_FC
+	bool "Bluetooth L2CAP flow control mode [EXPERIMENTAL]"
+	select EXPERIMENTAL
+	help
+	  This option enables Bluetooth L2CAP flow control mode
+
+config BT_L2CAP_ENH_RET
+	bool "Bluetooth L2CAP enhance retransmission [EXPERIMENTAL]"
+	select EXPERIMENTAL
+	help
+	  This option enables Bluetooth L2CAP enhance retransmission
+
+config BT_L2CAP_STREAM
+	bool "Bluetooth L2CAP streaming mode [EXPERIMENTAL]"
+	select EXPERIMENTAL
+	help
+	  This option enables Bluetooth L2CAP streaming mode
```
Are there really no interdependencies with these, i.e. all four of them can be enabled independently without depending on any other option?

lylezhu2012 replied:
Yes. I think they can be enabled independently.

lylezhu2012 commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -26,6 +27,84 @@ config BT_MAX_SCO_CONN
+	help
+	  This option enables Bluetooth L2CAP Extended window size Option
+
+config BT_L2CAP_MPS
+	int "Bluetooth L2CAP MPS for retransmission and Flow control"
+	default 48
+	range 48 BT_L2CAP_TX_MTU
+	help
+	  Bluetooth L2CAP MPS for retransmission and Flow control
+
+config BT_L2CAP_MAX_WINDOW_SIZE
+	int "Maximum Windows Size of Retransmission and Flow Control"
+	default 1
+	range 1 32
+	help
+	  Maximum Windows Size of Retransmission and Flow Control.
+	  The minimum (and default) number is 1.
+	  The range is 1 to 32 for Flow Control mode and Retransmission mode.
+	  The range is 1 to 63 for Enhanced Retransmission mode.
+	  The range is 1 to 0x3FFF for Enhanced Retransmission mode with
+	  externed control.
+
+config BT_L2CAP_BR_RET_TIMEOUT
+	int "Retransmission timeout"
+	default 2000
+	range 1000 2000
+	help
+	  the Retransmission timeout shall be three times the value of flush
+	  timeout, subject to a minimum of 1 second and maximum of 2 seconds.
+
+config BT_L2CAP_BR_MONITOR_TIMEOUT
+	int "Monitor timeout"
+	default 12000 if BT_L2CAP_ENH_RET
+	default BT_L2CAP_BR_RET_TIMEOUT
+	range BT_L2CAP_BR_RET_TIMEOUT 12000
```
They are depended on BT_L2CAP_RET, BT_L2CAP_FC, BT_L2CAP_ENH_RET, or BT_L2CAP_STREAM.

Let me update the `if` condition for them.

lylezhu2012 replied:
Updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -298,1706 +416,5208 @@ static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
+				&br_chan->_pdu_outstanding);
+			l2cap_br_free_window(br_chan, tx_win);
+			tx_win = (struct bt_l2cap_br_window *)sys_slist_peek_head(
+				&br_chan->_pdu_outstanding);
```
Updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -298,1706 +416,5208 @@ static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
-			connect_fixed_channel(BR_CHAN(chan));
+	LOG_DBG("chan %p retransmission timeout", br_chan);
+
+	if (sys_slist_peek_head(&br_chan->_pdu_outstanding)) {
```
Updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -298,1706 +416,5208 @@ static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
+			LOG_WRN("unacknowledged I-frame with sequence number %d",
+				br_chan->expected_ack_seq);
+			bt_l2cap_br_update_req_seq(br_chan, expected_ack_seq, false);
+		} else if (br_chan->tx.mode == BT_L2CAP_BR_LINK_MODE_ERET) {
```
Updated.

jhedberg commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -2412,6 +2412,11 @@ int bt_l2cap_chan_recv_complete(struct bt_l2cap_chan *chan, struct net_buf *buf)
 		return -ENOTCONN;
 	}
 
+	if (IS_ENABLED(CONFIG_BT_CLASSIC) &&
+		conn->type == BT_CONN_TYPE_BR) {
```
Incorrect split line alignment, however it also looks like the line split isn't needed at all.

lylezhu2012 replied:
Updated.

jhedberg commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -248,52 +329,321 @@ static const struct bt_l2cap_chan_ops l2cap_ops = {
+static int l2cap_accept(struct bt_conn *conn, struct bt_l2cap_server *server,
+			struct bt_l2cap_chan **chan);
+
+static struct bt_l2cap_br_server l2cap_server = {
+	.server = {.accept = l2cap_accept,},
+};
```
Just format this the same way you've done with `l2cap_chan` below, i.e. put the `.accept = l2cap_accept,` on its own line.

lylezhu2012 replied:
Updated.

jhedberg commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -248,52 +329,321 @@ static const struct bt_l2cap_chan_ops l2cap_ops = {
+#define BT_L2CAP_BR_SERVER_OPT_EXT_WIN_SIZE  BIT(5)
+#define BT_L2CAP_BR_SERVER_OPT_HOLD_CREDIT   BIT(6)
+
+static int l2cap_accept(struct bt_conn *conn, struct bt_l2cap_server *server,
+			struct bt_l2cap_chan **chan);
```
Is the forward declaration really necessary, i.e. do you have a circular dependency somewhere? If not, I'd consider reordering things to avoid it. Looking at the code before your PR it seems like the forward declaration wasn't necessary, but now for some reason you decided to move these before the `l2cap_accept()` function - why?

lylezhu2012 replied:
Move down `static struct bt_l2cap_br_server l2cap_server`. Remove the deceleration of `l2cap_accept()`.

jhedberg commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -2028,10 +5649,60 @@ void bt_l2cap_br_recv(struct bt_conn *conn, struct net_buf *buf)
+			 */
+			br_chan->retry_count = 1;
+
+			l2cap_br_start_timer(br_chan, false, true);
```
Unless the function takes just a single boolean parameter and it's clear from the name what that parameter does, having boolean parameters in functions is a terrible idea. Can you honestly say you'd know what the `false` and `true` do just by looking at this function call, if you didn't see the definition of the function?

lylezhu2012 replied:
Well.
I will change the type of second parameter to 
```
enum l2cap_br_timer_type {
	BT_L2CAP_BR_TIMER_RET,
	BT_L2CAP_BR_TIMER_MONITOR,
};
```
Then the function can be called likes as `l2cap_br_start_timer(br_chan, BT_L2CAP_BR_TIMER_MONITOR, true);`

jhedberg replied:
What does the third parameter do?

lylezhu2012 replied:
It is flag whether restart the timer.

jhedberg commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -299,1706 +417,5209 @@ static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
+		}
+	}))
+
+	if (br_chan->chan.ops->alloc_buf) {
+		err = bt_l2cap_br_recv_seg(br_chan, buf, sar);
+		if (err == -ENOBUFS) {
```
Just a small note about code flow: you seem to have a tendency to structure the functions so that you end up with some quite deep nesting. In many places you can avoid that by inverting conditions and doing early returns or gotos. E.g. in this case this could be:
```c
	if (err != -ENOBUFS) {
		goto done;
	}
```

There are many more places in your PR like this, but it's a huge PR so I won't go pointing them out individually.

lylezhu2012 replied:
The change can be useful for reducing one indentation for each line of code block. But It will introduce a additional `goto done;`. This change will cause doubts. From the perspective of reading, anti-logic is not conducive to reading.

lylezhu2012 replied:
> There are many more places in your PR like this, but it's a huge PR so I won't go pointing them out individually.

Sure I will check all changes and reduce the branch nesting.

jhedberg replied:
> anti-logic is not conducive to reading

I realize this is a pretty subjective topic, but it's a style that I and likely many others coming from the Linux kernel world are used to.

lylezhu2012 replied:
Well. It has been updated. And I have checked and fixed the similar issue.

jhedberg commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -299,1706 +417,5209 @@ static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
+				 */
+				discard = true;
+				LOG_DBG("Set REJ Exception");
+				if (!atomic_test_and_set_bit(br_chan->flags,
+							     L2CAP_FLAG_SEND_FRAME_REJ)) {
+					LOG_DBG("Set REJ Exception received flag");
```
Some really deep indentation here. It doesn't look like this might be solvable with the inverse conditions approach, so you might want to look at other kinds of restructuring, e.g. refactoring into helper functions.

lylezhu2012 replied:
Yes. It could be moved into a dedicated function. Let me update it.

lylezhu2012 replied:
Updated.

jhedberg reviewed (changes requested):
Some more review comments. This PR is so huge that I never have the time to look at the entirety of it in one go. This time I started from the bottom of the diff, to cover parts that I haven't looked at before, but I only got through maybe 1/5th of the PR.

jhedberg commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -26,6 +27,86 @@ config BT_MAX_SCO_CONN
+	help
+	  Maximum Windows Size of Retransmission and Flow Control.
+	  The minimum (and default) number is 1.
+	  The range is 1 to 32 for Flow Control mode and Retransmission mode.
+	  The range is 1 to 63 for Enhanced Retransmission mode.
+	  The range is 1 to 0x3FFF for Enhanced Retransmission mode with
```
This doesn't match the range definition on line 79, which always limits the value to max 32.

lylezhu2012 replied:
Since different protocols define different ranges, I took the parts they support in common. This way we can support different modes at runtime.

lylezhu2012 replied:
Removed these comments.

jhedberg commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -26,6 +27,86 @@ config BT_MAX_SCO_CONN
+	default 2000
+	range 1000 2000
+	help
+	  the Retransmission timeout shall be three times the value of flush
+	  timeout, subject to a minimum of 1 second and maximum of 2 seconds.
```
This doesn't explain that the value is in milliseconds. In fact, when you read the help text you get the impression that the value might be in seconds. If you don't need the millisecond-level granularity, maybe just specifying in seconds is sufficient?

lylezhu2012 replied:
Updated.

jhedberg commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -26,6 +27,86 @@ config BT_MAX_SCO_CONN
+	  monitor timeout shall be set to a value at least as large as the Retransmission
+	  timeout otherwise the value of the Monitor timeout shall be six times the value
+	  of flush timeout, subject to a minimum of the retransmission timeout value and
+	  a maximum of 12 seconds.
```
Same here.

lylezhu2012 replied:
Updated.

jhedberg commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -26,6 +27,86 @@ config BT_MAX_SCO_CONN
+	  externed control.
+
+config BT_L2CAP_BR_RET_TIMEOUT
+	int "Retransmission timeout"
```
"Retransmission timeout (milliseconds)" or "Retransmission timeout (seconds)" (depending on which unit you choose).

lylezhu2012 replied:
Updated.

jhedberg commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -32,75 +33,201 @@
+#define L2CAP_BR_PSM_END   0xffff
+
+#define L2CAP_BR_CID_DYN_START    0x0040
+#define L2CAP_BR_CID_DYN_END      0xffff
+#define L2CAP_BR_CID_IS_DYN(_cid) (_cid >= L2CAP_BR_CID_DYN_START && _cid <= L2CAP_BR_CID_DYN_END)
+
+#define L2CAP_BR_MIN_MTU     48
+#define L2CAP_BR_DEFAULT_MTU 672
+
+#define L2CAP_BR_PSM_SDP 0x0001
+
+#define L2CAP_BR_S_FRAME_FLAG_MASK 0xFFFFFF00U
```
Please decide if you want to use upper or lower case hex values. Now you're mixing both (lowercase on line 56 and upper case here). I've tried to use lower case everywhere in the Bluetooth host stack, but since there's so many contributors by now it's possible that there is some inconsistency. Please try to at least be consistent with yourself, or is there some reason why you're mixing these (e.g. special meaning with upper vs lower case?)

lylezhu2012 replied:
> Please decide if you want to use upper or lower case hex values. Now you're mixing both (lowercase on line 56 and upper case here). 

Line 56 is changed due to clang-format. 

> I've tried to use lower case everywhere in the Bluetooth host stack,

I did not notice that. Let me update the Line 64.

lylezhu2012 replied:
Updated all similar code lines.

jhedberg commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -36,16 +36,23 @@
 #endif /* CONFIG_BT_CONN */
 
-#define DATA_BREDR_MTU		48
+#if defined(CONFIG_BT_L2CAP_RET) || defined(CONFIG_BT_L2CAP_FC) ||                                 \
+	defined(CONFIG_BT_L2CAP_ENH_RET) || defined(CONFIG_BT_L2CAP_STREAM)
+#define L2CAP_BR_RET_FC_ENABLE 1
+#else
+#define L2CAP_BR_RET_FC_ENABLE 0
+#endif
```
This kind of construction sounds like those four Kconfig options should just select the same virtual option - then you don't need to do this additional stuff in the code itself.

lylezhu2012 replied:
Yes. Added a virtual config.

jhedberg commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -64,8 +71,10 @@ static int cmd_auth_pincode(const struct shell *sh,
-		shell_print(sh, "PIN code value invalid - enter max %u "
-			    "digits", max);
+		shell_print(sh,
+			    "PIN code value invalid - enter max %u "
+			    "digits",
```
This looks like it should fit on a single line

lylezhu2012 replied:
Let me check it. It is formatted by clang-format.

lylezhu2012 replied:
Updated.

jhedberg commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -312,8 +655,10 @@ static int cmd_discoverable(const struct shell *sh,
-		shell_print(sh, "BR/EDR set/reset discoverable failed "
-			    "(err %d)", err);
+		shell_print(sh,
+			    "BR/EDR set/reset discoverable failed "
+			    "(err %d)",
```
same here

lylezhu2012 replied:
Updated

jhedberg commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -381,30 +727,33 @@ static uint8_t sdp_hfp_ag_user(struct bt_conn *conn,
-			shell_error(ctx_shell, "Error getting Server CN, "
-				    "err %d", err);
+			shell_error(ctx_shell,
+				    "Error getting Server CN, "
+				    "err %d",
```
And here.

lylezhu2012 replied:
Updated.

jhedberg commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -415,15 +764,15 @@ static uint8_t sdp_hfp_ag_user(struct bt_conn *conn,
-			shell_error(ctx_shell, "Error getting HFPAG Features, "
-				    "err %d", err);
+			shell_error(ctx_shell,
+				    "Error getting HFPAG Features, "
+				    "err %d",
```
And here

lylezhu2012 replied:
Updated

jhedberg commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -441,36 +790,38 @@ static uint8_t sdp_a2src_user(struct bt_conn *conn,
-			    " remote %s", result->resp_buf,
-			    result->resp_buf->len, result->next_record_hint,
+		shell_print(ctx_shell,
+			    "SDP A2SRC data@%p (len %u) hint %u from"
+			    " remote %s",
```
and here

lylezhu2012 replied:
Updated

jhedberg commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -441,36 +790,38 @@ static uint8_t sdp_a2src_user(struct bt_conn *conn,
-			shell_error(ctx_shell, "A2SRC PSM Number not found, "
-				    "err %d", err);
+			shell_error(ctx_shell,
+				    "A2SRC PSM Number not found, "
+				    "err %d",
```
and here

lylezhu2012 replied:
Updated

jhedberg commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -441,36 +790,38 @@ static uint8_t sdp_a2src_user(struct bt_conn *conn,
-			shell_error(ctx_shell, "A2SRC version not found, "
-				    "err %d", err);
+			shell_error(ctx_shell,
+				    "A2SRC version not found, "
+				    "err %d",
```
and here

lylezhu2012 replied:
Updated

jhedberg commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -481,15 +832,15 @@ static uint8_t sdp_a2src_user(struct bt_conn *conn,
-			shell_error(ctx_shell, "A2SRC Features not found, "
-				    "err %d", err);
+			shell_error(ctx_shell,
+				    "A2SRC Features not found, "
+				    "err %d",
```
and here

lylezhu2012 replied:
Updated

jhedberg commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -293,1712 +405,5198 @@ static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
+	struct net_buf *sdu;
+
+	tx_win = (void *)sys_slist_peek_head(&br_chan->_pdu_outstanding);
+	while (tx_win) {
+		if (tx_win->tx_seq != req_seq) {
```
Wouldn't the loop be simpler (or at least have less indentation), if this was
```c
while (tx_win && tx_win->tx_seq != req_seq) {
```

lylezhu2012 replied:
Updated.

jhedberg commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -293,1712 +405,5198 @@ static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
+		break;
+	case BT_L2CAP_CONNECTING:
+		if (br_chan->state != BT_L2CAP_DISCONNECTED) {
+			LOG_WRN("%s()%d: invalid transition", func, line);
```
It's possible to configure logging so that function & line information is automatically included, i.e. it's redundant to do this as part of the format string itself.

jhedberg replied:
Ah, I see you're doing a similar trick here as the net_buf API, to catch the place where the higher level function was called, rather than the LOG_().

lylezhu2012 commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -32,75 +33,201 @@
+#define L2CAP_BR_RET_FC_ENABLE 1
+#else
+#define L2CAP_BR_RET_FC_ENABLE 0
+#endif
```
Added a config in Kconfig for it.

lylezhu2012 committed:
Bluetooth: BR: Shell: clang-format

Clang-format subsys/bluetooth/host/classic/shell/bredr.c

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

jhedberg commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -293,1712 +398,5196 @@ static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
+	return 0;
+}
+
+#if defined(CONFIG_BT_L2CAP_DYNAMIC_CHANNEL)
+#if defined(CONFIG_BT_L2CAP_LOG_LEVEL_DBG)
```
Just combine these to `#if defined(...) && defined(...)`. Then you also need just a single `#endif` at the end.

lylezhu2012 replied:
It is changed by clang-format. The change is reverted.

lylezhu2012 replied:
Updated here https://github.com/zephyrproject-rtos/zephyr/pull/86966

jhedberg commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -177,31 +177,31 @@ typedef enum bt_l2cap_chan_status {
+	uint16_t cid;
 	/** Endpoint Maximum Transmission Unit */
-	uint16_t				mtu;
+	uint16_t mtu;
 	/** Endpoint Maximum PDU payload Size */
-	uint16_t				mps;
+	uint16_t mps;
 	/** Endpoint credits */
-	atomic_t			credits;
+	atomic_t credits;
```
Just to be clear: clang-format is *not* a requirement for Zephyr coding style. It *can* be used for new code and should certainly result in code that's compatible with the Zephyr coding style. Here you're not using it for new code, so at the very least this style change should IMO be split out into a separate PR. Yes, consistency is important (e.g. same convention throughout the same c/h-file) but especially for big PRs like this it'd be important to keep the changes to an absolute minimum, i.e. don't include anything related to existing code.

lylezhu2012 replied:
I just push the new changes without doing clang-format for this file.

lylezhu2012 replied:
> Just to be clear: clang-format is _not_ a requirement for Zephyr coding style. It _can_ be used for new code and should certainly result in code that's compatible with the Zephyr coding style. Here you're not using it for new code, so at the very least this style change should IMO be split out into a separate PR. Yes, consistency is important (e.g. same convention throughout the same c/h-file) but especially for big PRs like this it'd be important to keep the changes to an absolute minimum, i.e. don't include anything related to existing code.

Well. But the output log of `./scripts/ci/check_compliance.py -c upstream/main..` is `You may want to run clang-format on this change`. 

This made me subconsciously use clang-format to format the file. There are a lot of changes in the format. I can only create a single commit to make such changes.

I will consider making a global format for classic code next. Is this reasonable?

lylezhu2012 commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -293,1712 +398,5196 @@ static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
+		/* do not care */
+		opt_qos->latency = sys_cpu_to_le32(0xffffffff);
+		/* do not care */
+		opt_qos->delay_variation = sys_cpu_to_le32(0xffffffff);
```
Updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -35,6 +36,85 @@ config BT_MAX_SCO_CONN
 	  supported. The minimum (and default) number is 1.
 
+config BT_L2CAP_RET_FC
+	def_bool BT_L2CAP_RET || BT_L2CAP_FC || BT_L2CAP_ENH_RET || BT_L2CAP_STREAM
```
Updated.

makeshi commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -299,1706 +410,5213 @@ static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
-		LOG_ERR("Too small L2CAP conn req packet size");
-		return;
+	if (!pdu_ready) {
+		LOG_DBG("nothing to send on this conn");
```
use LOG_ERR or LOG_WRN

makeshi commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -299,1706 +410,5213 @@ static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
+				/* No buffer can be used for transmission.
+				 * Waiting for buffer available.
+				 */
+				LOG_DBG("no buf for sending S-frame on %p", br_chan);
```
same above

lylezhu2012 replied:
Updated.

makeshi commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -299,1706 +410,5213 @@ static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
 
-static void l2cap_br_conf_req(struct bt_l2cap_br *l2cap, uint8_t ident,
-			      uint16_t len, struct net_buf *buf)
+static void l2cap_br_get_info(struct bt_l2cap_br *l2cap, uint16_t info_type)
```
It's more reasonable to use “send_info”. “get_info” seems to return information.

lylezhu2012 commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -271,42 +271,219 @@ struct bt_l2cap_le_chan {
+/** I-Frame transmission window for none `BASIC` mode L2cap connected channel. */
+struct bt_l2cap_br_window {
+	sys_snode_t node;
+
```
The space line is similar with this line
https://github.com/zephyrproject-rtos/zephyr/blob/210adaf399f47e4e72e79d1b0ace7e4fd124be23/include/zephyr/bluetooth/l2cap.h#L542-L545

lylezhu2012 commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -81,6 +188,23 @@ static sys_slist_t br_servers;
+#if defined(CONFIG_BT_L2CAP_RET_FC)
+static void br_tx_buf_destroy(struct net_buf *buf)
+{
+	net_buf_destroy(buf);
```
If the `buf` is `NULL`, the function `br_tx_buf_destroy` cannot be called. So, the `buf` should not be `NULL`.

lylezhu2012 commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -299,1706 +410,5213 @@ static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
-		LOG_DBG("last frag, removing %p", pdu);
-		__maybe_unused struct net_buf *b = k_fifo_get(&br_chan->_pdu_tx_queue, K_NO_WAIT);
+static void l2cap_br_sdu_is_done(struct bt_l2cap_br_chan *br_chan, struct net_buf *sdu, int err)
+{
```
If `l2cap_br_sdu_is_done` is renamed to `l2cap_br_is_sdu_done`, the return value could be `bool`.

lylezhu2012 commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -48,30 +55,130 @@ LOG_MODULE_REGISTER(bt_l2cap_br, CONFIG_BT_L2CAP_LOG_LEVEL);
+#define L2CAP_BR_CONN_TIMEOUT    K_SECONDS(40)
+
+#define L2CAP_FEAT_FC_MASK           BIT(0)
+#define L2CAP_FEAT_RET_MASK          BIT(1)
```
I refer to the spec. So I did not prefix/sub-fix.

![image](https://github.com/user-attachments/assets/bc3eafc5-5634-43ee-b569-576d74818577)

lylezhu2012 committed:
Bluetooth: L2CAP_BR: Sending multiple SDU at same time

Improve the retransmission and flow control to support sending
multiple SDU at the same time if the TX windows is not full.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -299,1706 +410,5213 @@ static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
-	STRUCT_SECTION_FOREACH(bt_l2cap_br_fixed_chan, fchan) {
-		struct bt_l2cap_br_chan *br_chan;
+	if (!ch) {
+		LOG_DBG("Unable to allocate L2CAP CID");
```
Updated.

lylezhu2012 commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -271,42 +271,219 @@ struct bt_l2cap_le_chan {
+	/** data address */
+	uint8_t *data;
+	/** data len */
+	uint16_t len;
```
Updated.

lylezhu2012 commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -271,42 +271,219 @@ struct bt_l2cap_le_chan {
+	 * `@kconfig{BT_L2CAP_MAX_WINDOW_SIZE}`. The field
+	 * should be no more then `CONFIG_BT_L2CAP_MAX_WINDOW_SIZE`.
+	 */
+	uint16_t                                window;
```
Updated.

makeshi reviewed (dismissed):
It is fine for me.

lylezhu2012 commented:
Rebase the latest

jhedberg commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -51,30 +58,130 @@ LOG_MODULE_REGISTER(bt_l2cap_br, CONFIG_BT_L2CAP_LOG_LEVEL);
+#define L2CAP_BR_S_RNR_FRAME            L2CAP_BR_S_FRAME_UD_FLAG_SET(BT_L2CAP_CONTROL_S_RNR)
+#define L2CAP_BR_S_SREJ_FRAME           L2CAP_BR_S_FRAME_UD_FLAG_SET(BT_L2CAP_CONTROL_S_SREJ)
+
+#define L2CAP_BR_INFO_TIMEOUT    K_SECONDS(4)
+#define L2CAP_BR_CFG_TIMEOUT     K_SECONDS(4)
+#define L2CAP_BR_DISCONN_TIMEOUT K_SECONDS(1)
+#define L2CAP_BR_CONN_TIMEOUT    K_SECONDS(40)
```
Do these values come from the spec? I'd assume so since they're not configurable, however it might make sense to have a code comment explaining their origin.

jhedberg replied:
I see this is just moved around from existing code, so not a blocker for this PR

lylezhu2012 replied:
The timer RTX is defined. `The value of this timer is implementation-dependent but the minimum initial value is
1 second and the maximum initial value is 60 seconds.`

jhedberg reviewed (approved):
I don't really have time to do a proper review of such huge PRs (even the GitHub web UI is super slow when trying to enter review comments). I also don't want to stand in the way of this going in, so I'll give my approval (just be aware that it doesn't mean that I've looked at every single line of code)
